refactor(common): use console service

LayZeeDK opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
`NgForOf` doesn't use the `Console` service and doesn't verify warnings in tests.

Issue Number: N/A


## What is the new behavior?
`NgForOf` uses the `Console` service and verifies warnings in tests.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

Adds a constructor argument of type `Console` to the `NgForOf` directive. Note that this is exported with the internal identifier `ɵConsole`.


## Other information

mhevery reviewed (commented):
Since console is only shown in dev mode, what is the advantage of this refactoring?

The downside of this change is that `Console` service is now retained whenever the `NgForOf` is retained.

mhevery commented:
There seem to be build errors which are failing the build, can you have a look at them?

LayZeeDK commented:
Thank you for your feedback. First, let's discuss whether it would be useful to use the Console service in all Angular core and common services an declarables or whether that's a bad idea. I see many different logging and testing strategies in the current codebase. What's your recommendation?

petebacondarwin commented:
I note that `Console` is already retained by `ApplicationRef` so adding it here would not make any difference. I note it is also used in a similar way in the `Router` class. So I think there is enough precedence here.

So I think this is a valid change and you should go ahead and rebase @LayZeeDK. Thanks for your patience.

On a more general note, if we are concerned about `Console` being retained. Would the following work? If instead of injecting the `Console` injectable, could we inject the `Injector`. And then, within the `ngDevMode` guard, use `this.injector.get(Console)`. Would that avoid `Console` being retained in production?

LayZeeDK committed:
refactor(common): use console service

Introduce the `Console` service to verify certain or no console
instructions during development.

Removes two long-standing TODOs.

LayZeeDK commented:
The mixed signals in the codebase that @petebacondarwin also noticed is exactly what I was asking about in my previous comment.

The initial trigger for this PR was to remove two long-standing TODOs from such an important part of the codebase. I explored the rest of the codebase and saw the many different approaches. The TODO comments in these files seemed to be waiting for something like the `Console` service which was added a long time ago.

LayZeeDK commented:
How can I fix this test from `ci/circleci: test` and `ci/circleci: test_ivy_aot`?

```
28 NgForOf.ɵfac = () => new NgForOfDef(
                        ~~~~~~~~~~~~~~~
29     ɵɵdirectiveInject(ViewContainerRef as any), ɵɵdirectiveInject(TemplateRef as any),
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
30     ɵɵdirectiveInject(IterableDiffers));
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  bazel-out/k8-fastbuild/bin/packages/common/src/directives/ng_for_of.d.ts:155:124
    155     constructor(_viewContainer: ViewContainerRef, _template: TemplateRef<NgForOfContext<T, U>>, _differs: IterableDiffers, _console: Console);
                                                                                                                                   ~~~~~~~~~~~~~~~~~
    An argument for '_console' was not provided.
```

Is it related to the golden file?

I'm unable to update the golden file. Every time I need to do anything with Bazel, I run into an issue on Windows.

```
$ C:\projects\angular\node_modules\.bin\bazel run --config=ivy //packages/core/test/bundling/todo:symbol_test.accept --verbose_failures
INFO: Analyzed target //packages/core/test/bundling/todo:symbol_test.accept (1 packages loaded, 66 targets configured).
INFO: Found 1 target...
ERROR: C:/projects/angular/packages/zone.js/lib/BUILD.bazel:9:11: Compiling TypeScript (devmode) //packages/zone.js/lib:zone_d_ts failed: Worker process quit or closed its stdin stream when we tried to send a WorkRequest:

---8<---8<--- Exception details ---8<---8<---
java.io.IOException: ERROR: src/main/native/windows/processes-jni.cc(356): NativeProcess:WriteStdin(15660): Pipen bliver lukket.

        at com.google.devtools.build.lib.windows.WindowsSubprocess.writeStream(WindowsSubprocess.java:263)
        at com.google.devtools.build.lib.windows.WindowsSubprocess.access$000(WindowsSubprocess.java:33)
        at com.google.devtools.build.lib.windows.WindowsSubprocess$ProcessOutputStream.write(WindowsSubprocess.java:57)
        at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.doFlush(CodedOutputStream.java:3062)
        at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.write(CodedOutputStream.java:2994)
        at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeLazy(CodedOutputStream.java:3013)
        at com.google.protobuf.ByteString$LiteralByteString.writeTo(ByteString.java:1392)
        at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeBytesNoTag(CodedOutputStream.java:2801)
        at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeBytes(CodedOutputStream.java:2775)
        at com.google.devtools.build.lib.worker.WorkerProtocol$Input.writeTo(WorkerProtocol.java:235)
        at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeMessageNoTag(CodedOutputStream.java:2855)
        at com.google.protobuf.CodedOutputStream$OutputStreamEncoder.writeMessage(CodedOutputStream.java:2824)
        at com.google.devtools.build.lib.worker.WorkerProtocol$WorkRequest.writeTo(WorkerProtocol.java:1081)
        at com.google.protobuf.AbstractMessageLite.writeDelimitedTo(AbstractMessageLite.java:95)
        at com.google.devtools.build.lib.worker.Worker.putRequest(Worker.java:127)
        at com.google.devtools.build.lib.worker.WorkerSpawnRunner.execInWorker(WorkerSpawnRunner.java:382)
        at com.google.devtools.build.lib.worker.WorkerSpawnRunner.actuallyExec(WorkerSpawnRunner.java:204)
        at com.google.devtools.build.lib.worker.WorkerSpawnRunner.exec(WorkerSpawnRunner.java:135)
        at com.google.devtools.build.lib.exec.SpawnRunner.execAsync(SpawnRunner.java:241)
        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:128)
        at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:96)
        at com.google.devtools.build.lib.actions.SpawnStrategy.beginExecution(SpawnStrategy.java:39)
        at com.google.devtools.build.lib.exec.SpawnStrategyResolver.beginExecution(SpawnStrategyResolver.java:62)
        at com.google.devtools.build.lib.analysis.actions.SpawnAction.beginExecution(SpawnAction.java:329)
        at com.google.devtools.build.lib.actions.Action.execute(Action.java:124)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$4.execute(SkyframeActionExecutor.java:781)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.continueAction(SkyframeActionExecutor.java:927)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:898)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:137)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:80)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:418)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:897)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:296)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:438)        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:398)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
---8<---8<--- End of exception details ---8<---8<---

---8<---8<--- Start of log, file at C:/users/lgbn/_bazel_lgbn/kxq2au6l/bazel-workers/worker-13-TypeScriptCompile.log ---8<---8<---
(empty)
---8<---8<--- End of log ---8<---8<---
Target //packages/core/test/bundling/todo:symbol_test.accept failed to build
INFO: Elapsed time: 1.241s, Critical Path: 0.22s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
error Command failed with exit code 1.
```

Can someone help me out here by running the `yarn bazel run --config=ivy //packages/core/test/bundling/todo:symbol_test.accept` command?

LayZeeDK commented:
Praise the WSL2 contributors 🙌

LayZeeDK commented:
How do I fix this failing test?

```
packages/core/test/render3/common_with_def.ts:28:22 - error TS2554: Expected 4 arguments, but got 3.

28 NgForOf.ɵfac = () => new NgForOfDef(
                        ~~~~~~~~~~~~~~~
29     ɵɵdirectiveInject(ViewContainerRef as any), ɵɵdirectiveInject(TemplateRef as any),
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
30     ɵɵdirectiveInject(IterableDiffers));
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  bazel-out/k8-fastbuild/bin/packages/common/src/directives/ng_for_of.d.ts:155:124
    155     constructor(_viewContainer: ViewContainerRef, _template: TemplateRef<NgForOfContext<T, U>>, _differs: IterableDiffers, _console: Console);
                                                                                                                                   ~~~~~~~~~~~~~~~~~
    An argument for '_console' was not provided.
```

Is it because of a golden file that needs to be updated except for the one I already updated (`packages/core/test/bundling/todo/bundle.golden_symbols.json`)? How do I do this?

JoostK commented:
@LayZeeDK if you open packages/core/test/render3/common_with_def.ts you should be able to change that handwritten factory function (line 28) to take an additional argument for the Console service.

LayZeeDK commented:
Since this is changing `common.d.ts`, I guess it is considered a breaking change?

JoostK commented:
> Since this is changing `common.d.ts` by adding an additional constructor parameter for `NgForOf`, is it considered a breaking change?

Technically yes. However, constructor parameters are considered an implementation detail and not officially part of the public API. Subclasses that override the constructor will definitely break, though, so this may need to target major in order to avoid the impact on apps.

LayZeeDK commented:
Alright, I'll change it in the PR description and figure out why this test is failing now.

LayZeeDK commented:
The `Console` service is internal. This is unfortunate for sub-classes extending `NgForOf`.

petebacondarwin commented:
@LayZeeDK if you still want to get this in, I would suggest changing the constructor to "optionally" inject the `Injector`. Then in the  dev mode guarded block we can then either use `Console` or fall back to `console`.

LayZeeDK commented:
I couldn't even figure out how to get the tests working again after rebasing. I don't think it's worth the effort at this point. Besides, it's just one corner of the codebase. A common strategy should be decided to normalize this across the codebase. Maybe it should be considered whether to expose the Console service in @angular/core or something.

petebacondarwin commented:
Sorry that this PR never got anywhere. I agree that we should have a consist approach to logging. I'll create an issue to track it.
